### ISSUE TITLE
Error in tasklist or titlebar widgets

### ISSUE THREAD

**Issue by guotsuan** [awesomeWM/awesome#521]
I found when I enable tasklist or titlebar widget, which will display the title information of clients.  I will get errors from time to time.  All display seems normal for me. And the errors happen only after awesome started for a while. 

The error information are

```
Error during delayed call: /usr/share/awesome/lib/wibox/drawable.lua:67: attempt to index a userdata value (field 'drawable')
stack traceback:
	/usr/share/awesome/lib/wibox/drawable.lua:67: in upvalue 'do_redraw'
	/usr/share/awesome/lib/wibox/drawable.lua:318: in field '?'
	/usr/share/awesome/lib/gears/timer.lua:151: in function </usr/share/awesome/lib/gears/timer.lua:150>
	[C]: in function 'xpcall'
```

And will repeat for multiple times. But all display are normal.  Like I said If I disable tasklist or titlebar widgets. The errors seems  vanish. I'm wondering if there is a good way to pin down the error. If anyone else ran into the similar problems.

the code in rc.lua for tasklist
```
for s in screen.count()
    ...
    mytasklist[s] = awful.widget.tasklist(s, awful.widget.tasklist.filter.currenttags, tasklist_buttons)
    ...
    local left_layout = wibox.layout.fixed.horizontal()
    left_layout:add(mylauncher)
    left_layout:add(mytaglist[s])
    left_layout:add(mypromptbox[s])
   ...
```
For float clients, I add titlebar to them with codes:
```
    local right_layout = wibox.layout.fixed.horizontal()
    right_layout:add(awful.titlebar.widget.floatingbutton(c))
    right_layout:add(awful.titlebar.widget.maximizedbutton(c))
    right_layout:add(awful.titlebar.widget.stickybutton(c))
    right_layout:add(awful.titlebar.widget.ontopbutton(c))
    right_layout:add(awful.titlebar.widget.closebutton(c))

    ---- The title goes in the middle
    --local title = awful.titlebar.widget.titlewidget(c)
    --title:buttons(awful.util.table.join(
            --awful.button({ }, 1, function()
                --client.focus = c
                --c:raise()
                --awful.mouse.client.move(c)
            --end),
            --awful.button({ }, 3, function()
                --client.focus = c
                --c:raise()
                --awful.mouse.client.resize(c)
            --end)
            --))

    ---- Now bring it all together
    local layout = wibox.layout.align.horizontal()
    --layout:set_left(left_layout)
    layout:set_right(right_layout)
    --layout:set_middle(title)

    awful.titlebar(c):set_widget(layout)
    awful.titlebar.show(c)
```
As shown above, If I comment out the ayout:set_middle(title) and related sentences. If enable them. I got errors.

**Comment by psychon** [comment]
What's your exact awesome version?

The symptom sounds similar to what I have been chasing for the last few days and reported yesterday as #520: Awesome tries redrawing a titlebar after it was already destroyed. (I got the error message ` attempt to index a userdata value (field 'drawable')` in my tests randomly, too.) You are lucky that your awesome version does not crash. :-)

The C code makes the drawin (a window you can draw to) unusable after it was garbage collected, but Lua allows getting access to finalized objects from finalizers and via tables with weak keys. Apparently we manage to get access to a finalized object via these weak tables.

**Comment by psychon** [comment]
Could you try something for you: Edit `lib/wibox/drawable.lua` and replace (around line 324, at least in current git/master):
```lua
            timer.delayed_call(ret._do_redraw)
            ret._redraw_pending = true
```
with
```lua
            ret._redraw_pending = true
            ret._do_redraw()
```
I hope that this does *not* make any difference, but the backtrace for the error should be different (Please provide the new backtrace). If this does make a difference, then something else is wrong...

**Comment by guotsuan** [comment]
My awesome version is:
awesome v3.5.2-805-g67faf52 (The Fox)
 • Build: Oct  9 2015 23:16:58 for x86_64 by gcc version 5.2.0 (dccf87@gqlinux)
 • Compiled against Lua 5.3.1 (running with Lua 5.3)
 • D-Bus support: ✔

Thanks for the suggestion. I will test it.

**Comment by guotsuan** [comment]
After I change the code. I got first two errors:
```
W: awesome: luaA_dofunction:77: error while running function
stack traceback:
	/usr/share/awesome/lib/wibox/drawable.lua:67: in upvalue 'do_redraw'
	/usr/share/awesome/lib/wibox/drawable.lua:318: in field '_do_redraw'
	/usr/share/awesome/lib/wibox/drawable.lua:325: in method 'draw'
	/usr/share/awesome/lib/wibox/drawable.lua:332: in local 'k'
	/usr/share/awesome/lib/wibox/drawable.lua:407: in function </usr/share/awesome/lib/wibox/drawable.lua:403>
error: /usr/share/awesome/lib/wibox/drawable.lua:67: attempt to index a userdata value (field 'drawable')
W: awesome: luaA_dofunction:77: error while running function
stack traceback:
	/usr/share/awesome/lib/wibox/drawable.lua:67: in upvalue 'do_redraw'
	/usr/share/awesome/lib/wibox/drawable.lua:318: in field '_do_redraw'
	/usr/share/awesome/lib/wibox/drawable.lua:325: in method 'draw'
	/usr/share/awesome/lib/wibox/drawable.lua:332: in local 'k'
	/usr/share/awesome/lib/wibox/drawable.lua:407: in function </usr/share/awesome/lib/wibox/drawable.lua:403>
error: /usr/share/awesome/lib/wibox/drawable.lua:67: attempt to index a userdata value (field 'drawable')
```
It seems related to the function
```
capi.awesome.connect_signal("wallpaper_changed", function()
    local k
    wallpaper = nil
    for k in pairs(drawables) do
        k()
    end
end)
```

I have a script to rotate my wall paper every 15 mins.
```
function wrotator.run (arg)
    local wp_path = arg.path  
    local timeout = arg.timeout or 600
    local filter = arg.filter or nil
    local wp_files = scandir(wp_path, filter)

    shuffleTable(wp_files)

    for s = 1, screen.count() do
        gears.wallpaper.maximized(wp_path .. wp_files[1], s)
        table.remove(wp_files, 1)
    end

    -- setup the timer
    wrotator.wp_timer = timer({timeout = timeout})

    wrotator.wp_timer:connect_signal("timeout", function()
     
      -- set wallpaper to current index for all screens
    if not next(wp_files) then
        wp_files = scandir(wp_path, filter)
        shuffleTable(wp_files)
    end

    for s = 1, screen.count() do
        while not file_exists(wp_path .. wp_files[1]) do
            table.remove(wp_files, 1)
        end

         gears.wallpaper.maximized(wp_path .. wp_files[1], s)
         table.remove(wp_files, 1)
    end

    --wp_timer 

    --if wp_timer.started then
      --wp_timer:stop()
    --else
      --wp_timer:start()
    --end

    end)

end
```
I'm still waiting for more errors. Previously, I can get 6 or 8 errors at the same time.

**Comment by psychon** [comment]
Ah, thanks. That explains how you can hit this problem of a drawin being repainted "while" getting destroyed.

**Comment by guotsuan** [comment]
I see. So my problem is due to the same reason as in  #520.

**Comment by psychon** [comment]
If you want to hit it more often, you could run `echo 'awesome.emit_signal("wallpaper_changed")' | awesome-client` a couple of times so that drawins get redrawn. But I don't think that will provide new insights. :-)

**Comment by guotsuan** [comment]
Thanks. And yes, every time I run ```echo 'awesome.emit_signal("wallpaper_changed")' |  awesome-client``` the same error will occur.

**Comment by psychon** [comment]
I have a new idea: `echo 'collectgarbage("collect")' | awesome-client`
If you do that (perhaps a couple of time?), can you still cause the error via the other account?

(Put differently: If you tell the garbage collector "please collect all garbage now", will it get rid of the half-destroyed titlebars and thus we stop trying to redraw them? I am asking because in my test case it seems that doing this changes the error, but doesn't make it go away. This is not what I expected it to do.)

**Comment by guotsuan** [comment]
Amazing.  After I run ```echo 'awesome.emit_signal("wallpaper_changed")' | awesome-client```.
Then ```echo 'awesome.emit_signal("wallpaper_changed")' | awesome-client``` would not cause any errors any more.  And even after my wallpaper was really changed, there was still no errors, at least for that time.

**Comment by psychon** [comment]
(I'll assume you meant `collectgarbage("collect")` for the first command. If not, say so, if yes then feel free to silently edit your comment and no one will notice. ;-) )
Thanks a lot. That supports my theory of "half destroyed wibox gets redrawn". I have to admit I half-doubted it after I didn't manage to come up with a simple test case nor a working fix.

**Comment by guotsuan** [comment]
:).  I mean collectgarbage("collect") for the first command.

**Comment by psychon** [comment]
Should be fixed by #567
